File: app/schemas/comprobantes.py
"""Esquemas de comprobantes.

> 🔑 **El comprobante no se emite: se registra.** El sistema legado no factura
> contra ARCA — alguien tipea el número de una factura que ya existe en papel o
> en el facturador de ARCA. Replicar eso es lo que da **paridad verificable**
> contra el sistema viejo durante la migración; la emisión real es F8, con su
> propio alcance. Mezclarlas haría que una diferencia de totales tuviera dos
> causas posibles y ninguna forma de separarlas.

> 🔑 **Los importes tampoco entran por el cuerpo: salen de las órdenes.** Un
> comprobante es la suma de las órdenes que agrupa. Aceptar un total del cliente
> permitiría que el comprobante diga un número y sus órdenes otro, que es
> exactamente la diferencia que el gate de F5 tiene que poder descartar.
"""

from datetime import date
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import TipoComprobante
from app.schemas.ordenes import OrdenOut

#: Los tipos que se registran sobre órdenes pendientes. Una nota de crédito no
#: agrupa órdenes: revierte un comprobante, y ese camino es `DELETE`.
TIPOS_FACTURA = frozenset({
    TipoComprobante.FACTURA_A,
    TipoComprobante.FACTURA_B,
    TipoComprobante.FACTURA_C,
})

#: Cómo se lee cada tipo en el concepto de la cuenta corriente.
NOMBRES_DE_TIPO = {
    TipoComprobante.FACTURA_A: "Factura A",
    TipoComprobante.FACTURA_B: "Factura B",
    TipoComprobante.FACTURA_C: "Factura C",
    TipoComprobante.NOTA_CREDITO_A: "Nota de credito A",
    TipoComprobante.NOTA_CREDITO_B: "Nota de credito B",
    TipoComprobante.NOTA_CREDITO_C: "Nota de credito C",
}


class FacturarIn(BaseModel):
    """"Facturar pendientes": las órdenes elegidas pasan a un comprobante."""

    model_config = ConfigDict(str_strip_whitespace=True)

    fecha: date
    razon_social_id: int
    cliente_id: int
    tipo: TipoComprobante
    punto_venta: int = Field(default=1, ge=0, le=99999)
    numero: int = Field(ge=1)
    orden_ids: list[int] = Field(min_length=1)

    @field_validator("orden_ids")
    @classmethod
    def _sin_repetidos(cls, valor: list[int]) -> list[int]:
        """Una orden repetida sumaría dos veces y la factura quedaría al doble.

        El `IN` de la consulta la trae una sola vez, así que sin este chequeo el
        pedido no falla: pasa, con un total que no es el de las órdenes.
        """
        if len(set(valor)) != len(valor):
            raise ValueError("hay ordenes repetidas: el importe se contaria dos veces")
        return valor


class ComprobanteOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    razon_social_id: int
    tipo: TipoComprobante
    punto_venta: int
    numero: int
    fecha: date
    cliente_id: int
    # Importes como `Decimal`: son `NUMERIC` en la base, y pasar por `float`
    # reintroduce el defecto que el producto viene a reparar.
    neto: Decimal
    iva: Decimal
    total: Decimal
    anulado: bool
    origen_legado: str | None = None


class SumaDeOrdenes(BaseModel):
    """Lo que suman las órdenes de un comprobante, contado aparte de él."""

    cantidad: int
    neto: Decimal
    iva: Decimal
    total: Decimal


class ComprobanteConOrdenes(BaseModel):
    """El comprobante, sus órdenes, y si los dos lados dan lo mismo.

    Es el gate de F5 a nivel de un comprobante: el encabezado guarda sus propios
    importes, y las órdenes los suyos. Que se devuelvan **los dos** y si
    coinciden evita tener que abrir la base para saber si el total del papel es
    el de las órdenes que lo componen.
    """

    comprobante: ComprobanteOut
    ordenes: list[OrdenOut]
    suma_de_ordenes: SumaDeOrdenes
    coinciden: bool


class TotalDeRazonSocial(BaseModel):
    """El total facturado por una razón social, contado por los dos lados.

    - Por **comprobantes**: suma de los encabezados.
    - Por **órdenes**: suma de las órdenes, agrupadas por la razón social que
      lleva **la orden**, no la del comprobante.

    Agrupar por la columna de la orden es a propósito: `facturar` deja las dos
    iguales, así que sobre datos cargados acá siempre coinciden. Lo que esto
    detecta son los **datos migrados**, donde `carga_razonsocial` y
    `factura_razonsocial` son dos columnas del legado que pueden discrepar — y
    entonces el mismo importe estaría en una razón social por un lado y en otra
    por el otro.
    """

    razon_social_id: int | None
    cantidad_comprobantes: int
    neto_comprobantes: Decimal
    iva_comprobantes: Decimal
    total_comprobantes: Decimal
    cantidad_ordenes: int
    neto_ordenes: Decimal
    iva_ordenes: Decimal
    total_ordenes: Decimal
    coinciden: bool

File: frontend/src/api/comprobantes.ts
import { api } from 'libra-ui/api-client'

import type { Orden } from '@/api/ordenes'

export type TipoComprobante =
  | 'factura_a' | 'factura_b' | 'factura_c'
  | 'nota_credito_a' | 'nota_credito_b' | 'nota_credito_c'

/** Lo que devuelve facturar cuando el ambiente de ARCA es homologación.
 *
 *  🔴 **No es un `Comprobante` incompleto: no existe.** El backend corre el
 *  alta entera contra ARCA —número, pedido, CAE— y la revierte, porque acá un
 *  comprobante además mueve la cuenta corriente y cierra las órdenes. Por eso
 *  no tiene `id`, y por eso el `POST` contesta 200 y no 201.
 *
 *  Se distingue por `ensayo`, que sólo viene en esta forma. Guiarse por la
 *  ausencia de `id` sería frágil: cualquier respuesta a medias la cumpliría.
 */
export type Ensayo = {
  ensayo: true
  ambiente: string
  tipo: TipoComprobante
  punto_venta: number
  numero: number
  total: string
  cae: string | null
  cae_vencimiento: string | null
}

export type Comprobante = {
  id: number
  razon_social_id: number
  tipo: TipoComprobante
  punto_venta: number
  numero: number
  fecha: string
  cliente_id: number
  // Importes como STRING, igual que en las ordenes: son `NUMERIC` en la base y
  // `Decimal` en Python. Pasarlos por `number` los mete en un float binario,
  // que es el defecto que el producto viene a reparar.
  neto: string
  iva: string
  total: string
  anulado: boolean
  origen_legado: string | null
}

export type SumaDeOrdenes = { cantidad: number; neto: string; iva: string; total: string }

export type ComprobanteConOrdenes = {
  comprobante: Comprobante
  ordenes: Orden[]
  suma_de_ordenes: SumaDeOrdenes
  /** Si el encabezado dice lo mismo que sus ordenes. */
  coinciden: boolean
}

export type TotalDeRazonSocial = {
  razon_social_id: number | null
  cantidad_comprobantes: number
  neto_comprobantes: string
  iva_comprobantes: string
  total_comprobantes: string
  cantidad_ordenes: number
  neto_ordenes: string
  iva_ordenes: string
  total_ordenes: string
  coinciden: boolean
}

/** Un importe `"1234.56"` a centavos enteros. */
function aCentavos(valor: string): number {
  const [entero, decimales = ''] = valor.trim().split('.')
  const signo = entero.trimStart().startsWith('-') ? -1 : 1
  return signo * (Math.abs(Number(entero)) * 100 + Number((decimales + '00').slice(0, 2)))
}

/** Suma importes **en centavos enteros**, no en punto flotante.
 *
 * `0.1 + 0.2` en JavaScript da `0.30000000000000004`: los importes se manejan
 * como texto en toda la app justamente para no pasar por ahi. Esta suma es solo
 * la **previsualizacion** de lo que se va a facturar --el importe que queda
 * guardado lo calcula el servidor con `Decimal` sobre las mismas ordenes--,
 * pero una vista previa que no coincide con el total real es peor que no
 * mostrar nada: quien la mira decide con ella.
 */
export function sumarImportes(valores: string[]): string {
  const centavos = valores.reduce((acumulado, v) => acumulado + aCentavos(v), 0)
  const signo = centavos < 0 ? '-' : ''
  const absoluto = Math.abs(centavos)
  return `${signo}${Math.floor(absoluto / 100)}.${String(absoluto % 100).padStart(2, '0')}`
}

/** `Factura A 0001-00000123`, como se lee en el papel. */
export const NOMBRE_DE_TIPO: Record<TipoComprobante, string> = {
  factura_a: 'Factura A', factura_b: 'Factura B', factura_c: 'Factura C',
  nota_credito_a: 'Nota de crédito A', nota_credito_b: 'Nota de crédito B',
  nota_credito_c: 'Nota de crédito C',
}

export function numeroDe(c: Comprobante): string {
  return `${String(c.punto_venta).padStart(4, '0')}-${String(c.numero).padStart(8, '0')}`
}

export const comprobantes = {
  listar: (filtros: Record<string, string | number | boolean | undefined> = {}) => {
    const p = new URLSearchParams()
    // `!= null` y no `if (v)`: `anulado=false` es un filtro y no una ausencia.
    for (const [k, v] of Object.entries(filtros)) {
      if (v != null && v !== '') p.set(k, String(v))
    }
    const qs = p.toString()
    return api.get<Comprobante[]>(`/api/comprobantes${qs ? `?${qs}` : ''}`)
  },
  ver: (id: number) => api.get<ComprobanteConOrdenes>(`/api/comprobantes/${id}`),
  totales: (desde?: string, hasta?: string) => {
    const p = new URLSearchParams()
    if (desde) p.set('desde', desde)
    if (hasta) p.set('hasta', hasta)
    const qs = p.toString()
    return api.get<TotalDeRazonSocial[]>(`/api/comprobantes/totales${qs ? `?${qs}` : ''}`)
  },
  // 🔑 La unión no es cosmética: obliga a quien llame a **decidir cuál de
  // los dos recibió** antes de tocar `.id`. Con `Comprobante` a secas,
  // TypeScript deja leer el id de un ensayo —que no lo tiene— y el error
  // aparece recién en pantalla, como una navegación a la nada.
  facturar: (datos: unknown) =>
    api.post<Comprobante | Ensayo>('/api/comprobantes', datos),
  anular: (id: number) => api.del<Comprobante>(`/api/comprobantes/${id}`),
}
